[jsc] Add option to force the address of JIT memory

Achierius opened the pull request:
#### c42cc0d597b192b5c92ba019aa8167cc26aca31d
<pre>
[jsc] Add option to force the address of JIT memory
<a href="https://bugs.webkit.org/show_bug.cgi?id=297560">https://bugs.webkit.org/show_bug.cgi?id=297560</a>
<a href="https://rdar.apple.com/158636737">rdar://158636737</a>

Reviewed by Yusuke Suzuki and Mark Lam.

The new JSC_jitMemoryReservationAddress option allows us to specify an
address at which to mmap JIT memory. If the operating system is not able
to do so, then we will instead crash.
Notably we do not use MAP_FIXED as we don&apos;t want to clobber existing
mappings, but would rather bail out if someone else has already mapped
the location requested.

Canonical link: <a href="https://commits.webkit.org/299073@main">https://commits.webkit.org/299073@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f91004c6ca986e84f63c2ce24b64d3103a7a513

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117700 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37378 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28006 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123834 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/69713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b73179c9-8fcb-4e39-b018-a62e6b12254c) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45960 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/89331 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/69713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c1b2ca18-54d4-419e-9763-a370b29793f0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120652 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/30313 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/105540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69823 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/75beb81d-6196-4906-8c9d-18b4eb161a14) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/29376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/23656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67492 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/109811 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/99731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/23836 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126927 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/116209 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44603 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33570 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/97995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44961 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101767 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/97782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24882 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43154 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/21123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44475 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/50149 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144907 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43933 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/37289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/47280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45624 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e1ed7caba33c65f6702923d099bcfff68621a7c1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49563

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/21895f92b0080f027e3e57e172f681fedbb2eafa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49563

Constellation commented on Source/WebKit/Platform/IPC/Encoder.cpp:
```diff
@@ -45,7 +45,7 @@ static constexpr uint8_t defaultMessageFlags = 0;
 static inline MallocSpan<uint8_t, Mmap> allocateBuffer(size_t size)
 {
-    auto buffer = MallocSpan<uint8_t, Mmap>::mmap(size, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1);
+    auto buffer = MallocSpan<uint8_t, Mmap>::mmap(0, size, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1);
```
Maybe using nullptr?

Constellation commented on Source/WTF/wtf/OSAllocator.h:
```diff
@@ -99,7 +99,7 @@ inline void OSAllocator::decommitAndRelease(void* releaseBase, size_t releaseSiz
 inline T* OSAllocator::reallocateCommitted(T* oldBase, size_t oldSize, size_t newSize, Usage usage, bool writable, bool executable, bool jitCageEnabled)
 {
-    void* newBase = reserveAndCommit(newSize, usage, writable, executable, jitCageEnabled);
+    void* newBase = reserveAndCommit(newSize, usage, 0, writable, executable, jitCageEnabled);
```
Ditto, using `nullptr`.

Constellation commented on Source/WTF/wtf/PageAllocation.h:
```diff
@@ -84,7 +84,7 @@ class PageAllocation : private PageBlock {
     {
         ASSERT(isPageAligned(size));
-        return PageAllocation(OSAllocator::reserveAndCommit(size, usage, writable, executable), size);
+        return PageAllocation(OSAllocator::reserveAndCommit(size, usage, 0, writable, executable), size);
```
Ditto.

Constellation commented on Source/WTF/wtf/PageReservation.h:
```diff
@@ -89,41 +89,48 @@ class PageReservation : private PageBlock {
-        return PageReservation(OSAllocator::reserveUncommitted(size, usage, writable, executable, jitCageEnabled, false), size, writable, executable, jitCageEnabled, false);
+        ASSERT(isPageAligned(address.value_or(0)));
+        if (commit)
+            return PageReservation(OSAllocator::reserveAndCommit(size, usage, (void*)address.value_or(0),             writable, executable, jitCageEnabled, false), size, writable, executable, jitCageEnabled, false);
+        return PageReservation(OSAllocator::reserveUncommitted(size, usage, (void*)address.value_or(0),             writable, executable, jitCageEnabled, false), size, writable, executable, jitCageEnabled, false);
```
We should remove spaces between `value_or(0),` and `writable`. Also, let's use `std::bit_cast<void*>(...)`.

Constellation commented on Source/WTF/wtf/PageReservation.h:
```diff
@@ -89,41 +89,48 @@ class PageReservation : private PageBlock {
-        return PageReservation(OSAllocator::tryReserveUncommitted(size, usage, writable, executable, jitCageEnabled, false), size, writable, executable, jitCageEnabled, false);
+        ASSERT(isPageAligned(address.value_or(0)));
+        if (commit)
+            return PageReservation(OSAllocator::tryReserveAndCommit(size, usage, (void*)address.value_or(0),             writable, executable, jitCageEnabled, false), size, writable, executable, jitCageEnabled, false);
+        return PageReservation(OSAllocator::tryReserveUncommitted(size, usage, (void*)address.value_or(0),             writable, executable, jitCageEnabled, false), size, writable, executable, jitCageEnabled, false);
```
Ditto.

Constellation commented on Source/WTF/wtf/PageReservation.h:
```diff
@@ -89,41 +89,48 @@ class PageReservation : private PageBlock {
+        if (address.has_value() && address.value() >= pageSize())
+            address.value() -= pageSize();
+        if (commit)
+            return PageReservation(OSAllocator::reserveAndCommit(size + pageSize() * 2, usage, (void*)address.value_or(0),             writable, executable, jitCageEnabled, true), size, writable, executable, jitCageEnabled, true);
+        return PageReservation(OSAllocator::reserveUncommitted(size + pageSize() * 2, usage, (void*)address.value_or(0), writable, executable, jitCageEnabled, true), size, writable, executable, jitCageEnabled, true);
```
Ditto.

Constellation commented on Source/WTF/wtf/PageReservation.h:
```diff
@@ -89,41 +89,48 @@ class PageReservation : private PageBlock {
+        if (address.has_value() && address.value() >= pageSize())
+            address.value() -= pageSize();
+        if (commit)
+            return PageReservation(OSAllocator::tryReserveAndCommit(size + pageSize() * 2, usage, (void*)address.value_or(0),             writable, executable, jitCageEnabled, true), size, writable, executable, jitCageEnabled, true);
+        return PageReservation(OSAllocator::tryReserveUncommitted(size + pageSize() * 2, usage, (void*)address.value_or(0),             writable, executable, jitCageEnabled, true), size, writable, executable, jitCageEnabled, true);
```
Ditto.

Achierius committed:
[jsc] Add option to force the address of JIT memory
https://bugs.webkit.org/show_bug.cgi?id=297560
rdar://158636737

Reviewed by Yusuke Suzuki and Mark Lam.

The new JSC_jitMemoryReservationAddress option allows us to specify an
address at which to mmap JIT memory. If the operating system is not able
to do so, then we will instead crash.
Notably we do not use MAP_FIXED as we don't want to clobber existing
mappings, but would rather bail out if someone else has already mapped
the location requested.

Canonical link: https://commits.webkit.org/299073@main
